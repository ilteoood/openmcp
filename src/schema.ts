import { type ZodRawShape, z } from "zod";
import type { ParamRequestObject, ParameterObject } from "./types.js";

// biome-ignore lint/suspicious/noExplicitAny: this is the definition of the schema from the library
type Schema = any;

const buildBaseZodType = (schema: Schema) => {
	if (schema.type === "string") {
		if (schema.enum) {
			return z.enum(schema.enum);
		}
		return z.string();
	}
	if (schema.type === "number" || schema.type === 'integer') {
		let integerSchema = z.number();
		if (schema.minimum !== undefined) {
			integerSchema = integerSchema.min(schema.minimum);
		}
		if (schema.maximum !== undefined) {
			integerSchema = integerSchema.max(schema.maximum);
		}
		if (schema.type === 'integer') {
			integerSchema = integerSchema.int();
		}
		return integerSchema;
	}
	if (schema.type === "boolean") {
		return z.boolean();
	}
	if (schema.type === "array") {
		return z.array(createFromSchema(schema.items));
	}
	if (schema.type === "object") {
		const zodObjectShape = Object.entries<Schema>(
			schema.properties || {},
		).reduce<ParamRequestObject>((acc, [propertyName, property]) => {
			let zodSchema = createFromSchema(property);
			if (!property.required && !schema.required?.includes(propertyName)) {
				zodSchema = zodSchema.optional();
			}

			acc[propertyName] = zodSchema;
			return acc;
		}, {});
		return z.object(zodObjectShape);
	}
	return z.any();
};

export const createFromSchema = (
	schema: Schema,
	parameter?: ParameterObject,
): z.ZodType => {
	const description = parameter?.description || schema?.description || "";
	const zodType = buildBaseZodType(schema).describe(description);

	if (!parameter?.required && !schema?.required) {
		return zodType.optional();
	}

	return zodType;
};

export const createFromParameter = (parameter: ParameterObject): z.ZodType =>
	createFromSchema(parameter.schema, parameter);

export const createForObject = <T extends ZodRawShape>(object?: T) =>
	object && Object.keys(object).length ? z.object(object) : z.void();

export const createFromParameters = (parameterObjects: ParameterObject[]) => {
	const parameters = parameterObjects.reduce<ParamRequestObject>(
		(acc, parameter) => {
			const zodSchema = createFromParameter(parameter);
			acc[parameter.name] = zodSchema;
			return acc;
		},
		{},
	);

	return createForObject(parameters);
};
