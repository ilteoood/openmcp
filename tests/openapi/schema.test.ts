import { describe, expect, it } from "vitest";
import { createFromParameter } from "../../src/schema.js";
import { loadFixtures } from "../utils.js";

describe("schema", () => {
	const fieldsToGenerate = [
		{ type: "integer", index: 0 },
		{ type: "string", index: 1 },
		{ type: "boolean", index: 2 },
		{ type: "number", index: 3 },
		{ type: "array", index: 4 },
		{ type: "object", index: 5 },
		{ type: "integer not required", index: 6 },
	];

	it.each(fieldsToGenerate)(
		"should create a zod schema for a $type type",
		async ({ type, index }) => {
			const petSchema = await loadFixtures("schema.json");

			const petIdParameter =
				petSchema.paths["/pet/{petId}"].post.parameters[index];

			const zodSchema = createFromParameter(petIdParameter);

			expect(zodSchema).toMatchSnapshot();
		},
	);
});
