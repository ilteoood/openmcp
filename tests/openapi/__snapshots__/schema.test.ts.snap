// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema > should create a zod schema for a 'array' type 1`] = `
ZodOptional {
  "_def": {
    "description": "List of owners of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodArray {
      "_def": {
        "description": "List of owners of the pet that needs to be updated",
        "exactLength": null,
        "maxLength": null,
        "minLength": null,
        "type": ZodOptional {
          "_def": {
            "description": "",
            "errorMap": [Function],
            "innerType": ZodNumber {
              "_def": {
                "checks": [
                  {
                    "kind": "int",
                    "message": undefined,
                  },
                ],
                "coerce": false,
                "description": "",
                "typeName": "ZodNumber",
              },
              "and": [Function],
              "array": [Function],
              "brand": [Function],
              "catch": [Function],
              "default": [Function],
              "describe": [Function],
              "isNullable": [Function],
              "isOptional": [Function],
              "max": [Function],
              "min": [Function],
              "nullable": [Function],
              "nullish": [Function],
              "optional": [Function],
              "or": [Function],
              "parse": [Function],
              "parseAsync": [Function],
              "pipe": [Function],
              "promise": [Function],
              "readonly": [Function],
              "refine": [Function],
              "refinement": [Function],
              "safeParse": [Function],
              "safeParseAsync": [Function],
              "spa": [Function],
              "step": [Function],
              "superRefine": [Function],
              "transform": [Function],
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
            "typeName": "ZodOptional",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "typeName": "ZodArray",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'boolean' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Status of pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodBoolean {
      "_def": {
        "coerce": false,
        "description": "Status of pet that needs to be updated",
        "typeName": "ZodBoolean",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'integer not required' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Age of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodNumber {
      "_def": {
        "checks": [
          {
            "kind": "int",
            "message": undefined,
          },
        ],
        "coerce": false,
        "description": "Age of the pet that needs to be updated",
        "typeName": "ZodNumber",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "max": [Function],
      "min": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "step": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'integer with min and max' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Age of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodNumber {
      "_def": {
        "checks": [
          {
            "inclusive": true,
            "kind": "min",
            "message": undefined,
            "value": 10,
          },
          {
            "inclusive": true,
            "kind": "max",
            "message": undefined,
            "value": 40,
          },
          {
            "kind": "int",
            "message": undefined,
          },
        ],
        "coerce": false,
        "description": "Age of the pet that needs to be updated",
        "typeName": "ZodNumber",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "max": [Function],
      "min": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "step": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'integer' type 1`] = `
ZodNumber {
  "_def": {
    "checks": [
      {
        "kind": "int",
        "message": undefined,
      },
    ],
    "coerce": false,
    "description": "ID of pet that needs to be updated",
    "typeName": "ZodNumber",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "max": [Function],
  "min": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "step": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'number with min and max' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Width of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodNumber {
      "_def": {
        "checks": [
          {
            "inclusive": true,
            "kind": "min",
            "message": undefined,
            "value": 0,
          },
          {
            "inclusive": true,
            "kind": "max",
            "message": undefined,
            "value": 100,
          },
        ],
        "coerce": false,
        "description": "Width of the pet that needs to be updated",
        "typeName": "ZodNumber",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "max": [Function],
      "min": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "step": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'number' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Height of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodNumber {
      "_def": {
        "checks": [],
        "coerce": false,
        "description": "Height of the pet that needs to be updated",
        "typeName": "ZodNumber",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "max": [Function],
      "min": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "step": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'object' type 1`] = `
ZodOptional {
  "_def": {
    "description": "General informations about the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "description": "General informations about the pet that needs to be updated",
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'string with enum' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Status of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodEnum {
      "_def": {
        "description": "Status of the pet that needs to be updated",
        "typeName": "ZodEnum",
        "values": [
          "available",
          "pending",
          "sold",
        ],
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'string with minLength and maxLength' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Profile picture of the pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodString {
      "_def": {
        "checks": [
          {
            "kind": "min",
            "value": 3,
          },
          {
            "kind": "max",
            "value": 300,
          },
        ],
        "coerce": false,
        "description": "Profile picture of the pet that needs to be updated",
        "typeName": "ZodString",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`schema > should create a zod schema for a 'string' type 1`] = `
ZodOptional {
  "_def": {
    "description": "Name of pet that needs to be updated",
    "errorMap": [Function],
    "innerType": ZodString {
      "_def": {
        "checks": [],
        "coerce": false,
        "description": "Name of pet that needs to be updated",
        "typeName": "ZodString",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "typeName": "ZodOptional",
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;
